core/src/main/java/aima/core/search/basic/support/BasicProblem.java

```diff
@@ -102,4 +114,14 @@ public boolean isGoalState(S state) {
 	public double stepCost(S s, A a, S sPrime) {
 		return stepCostFn.stepCost(s, a, sPrime);
 	}
+
+	@Override
+	public Problem<A, S> getOriginalProblem() {
+		return this;
+	}
+
+	@Override
+	public Problem<A, S> getReverseProblem() {
+		return new BasicProblem<A, S>(finalState, initialState, goalTestPredicate, actionsFn, resultFn, stepCostFn);
```

Comment:
More efficient to instantiate just once and keep as an attribute (see earlier comments).